Fix Web Inspector: Remember the message type selection in the Console tab

the-chenergy opened the pull request:
#### 08a1f3998383fe5de26c9910bff87e89cb1739f3
<pre>
Fix Web Inspector: Remember the message type selection in the Console tab
<a href="https://rdar.apple.com/122924275">rdar://122924275</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=268882">https://bugs.webkit.org/show_bug.cgi?id=268882</a>

Reviewed by Devin Rousso.

When showing the the inspector&apos;s console using `WI.showConsole()`, the
caller can optionally pass in a `requestedScope` to control which levels
(AKA message types) to be filtered automatically when the Console tab
shows up. However, when `requestedScope` is falsy or left empty, it
always applies `WI.LogContentView.Scopes.All` instead, which
overrides the levels selected by default, which are read from local
settings when the scope bar is created at the inspector&apos;s startup.

This commit removes the forced application of `Scopes.All`, so when
`requestedScope` is left empty, the Console tab is shown with levels
unchanged, which is the expected behavior when launching the Console tab
through Develop -&gt; Show JavaScript Console (or Option-Command-C).

This fix has one known side-effect: when an inspector tab does not
support split console view, pressing Esc will switch to the actual
Console tab instead. (The Settings tab is one example of such tab.)
This commit will make that also &quot;remember&quot; the previously selected
levels instead of deselecting back to just `Scopes.All`, which is
arguably the correct behavior anyway.

This commit also cleans up on how `requestedScope` gets passed in;
passing in as part of the `options` parameter allows callers of
`showConsole()` to self-document the usage `requestedScope`.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
  - Fix the bug.
  - Adapt to the clean up for the `options` parameter.
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.showConsole):
  - Adapt to the clean up for the `options` parameter.
* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._showConsoleTab):
  - Adapt to the clean up for the `options` parameter.

Canonical link: <a href="https://commits.webkit.org/275143@main">https://commits.webkit.org/275143@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0a4a0ec520016237521577c7cab7cfb76ec287ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40988 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43366 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43547 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37080 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23011 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17332 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33948 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41562 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16921 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14575 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14686 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44850 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37175 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38737 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17422 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9199 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/17473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ac3f308f5a705c9c91891260f10fcb6469730eb0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24831

rcaliman-apple commented:
Had this conversation offline with @the-chenergy, but repeating it here for visibility to address confusion I had and others might have too:

The saved state of each scope bar item is pulled from a corresponding `WI.Setting` when creating its view:
https://github.com/WebKit/WebKit/blob/fb47415c1ad4bfc008367fffe637384f93757531/Source/WebInspectorUI/UserInterface/Views/ScopeBarItem.js#L48-L49

This patch allows that behaviour to run when `requestedScope` is `null` because of `InspectorFrontendApi.showConsole()`:
https://github.com/WebKit/WebKit/blob/fb47415c1ad4bfc008367fffe637384f93757531/Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js#L95-L100

rcaliman-apple reviewed (approved):
We generally don't add/remove whitespace unrelated to the patch itself. 
This is correcting some leftover tabs. I guess that's ok, @dcrousso?

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -1201,11 +1201,10 @@ WI.hideSplitConsole = function()
     WI.hideSplitConsole();
 
-    WI.consoleContentView.scopeBar.item(requestedScope).selected = true;
+    if (requestedScope)
```
NIT: if you really wanted to be ideal we'd move `requestedScope` to be inside `options` so that callers don't have do provide it if they don't need it (e.g. no need to have a `const requestedScope = null;`)
```js
if (options.requestedScope)
    WI.consoleContentView.scopeBar.item(options.requestedScope).selected = true;
```

rcaliman-apple replied:
> While we're at this though, since the Console tab technically supports enabling multiple levels at once (e.g. Errors and Warnings), should I go ahead and change `requestedScope` to `requestedScopes` and make it take an array of scopes instead? Or do you think that'd be too many tangential changes to this commit was meant to do?
> 
> CC @rcaliman-apple

The only things that currently trigger the codepath to show Console with a specific scope are the errors and warnings buttons in the Tab bar:

https://github.com/WebKit/WebKit/blob/1c84cd6c970271eabe238a7e4eff9adced5614ce/Source/WebInspectorUI/UserInterface/Base/Main.js#L455-L468

Since there's no caller with multiple scopes, I don't think it's necessary to convert `requestedScope` to an array now. We can do it whenever it's needed.

dcrousso reviewed (approved):
> We generally don't add/remove whitespace unrelated to the patch itself. This is correcting some leftover tabs. I guess that's ok, @dcrousso?

if the whitespace was nearby the relevant changes, id be ok with it, but as-is they're pretty far away so id prefer if they were not included in the same change

webkit-commit-queue commented:
@the-chenergy does not have committer permissions according to https://raw.githubusercontent.com/WebKit/WebKit/main/metadata/contributors.json.

If you do have committer permmissions, please ensure that your GitHub username is added to contributors.json.

Rejecting ac3f308f5a705c9c91891260f10fcb6469730eb0 from merge queue.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ae0870f7123604d52174c3091b12e76b3fee33f4. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24831

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0a4a0ec520016237521577c7cab7cfb76ec287ba. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/24831

the-chenergy committed:
Fix Web Inspector: Remember the message type selection in the Console tab
rdar://122924275
https://bugs.webkit.org/show_bug.cgi?id=268882

Reviewed by Devin Rousso.

When showing the the inspector's console using `WI.showConsole()`, the
caller can optionally pass in a `requestedScope` to control which levels
(AKA message types) to be filtered automatically when the Console tab
shows up. However, when `requestedScope` is falsy or left empty, it
always applies `WI.LogContentView.Scopes.All` instead, which
overrides the levels selected by default, which are read from local
settings when the scope bar is created at the inspector's startup.

This commit removes the forced application of `Scopes.All`, so when
`requestedScope` is left empty, the Console tab is shown with levels
unchanged, which is the expected behavior when launching the Console tab
through Develop -> Show JavaScript Console (or Option-Command-C).

This fix has one known side-effect: when an inspector tab does not
support split console view, pressing Esc will switch to the actual
Console tab instead. (The Settings tab is one example of such tab.)
This commit will make that also "remember" the previously selected
levels instead of deselecting back to just `Scopes.All`, which is
arguably the correct behavior anyway.

This commit also cleans up on how `requestedScope` gets passed in;
passing in as part of the `options` parameter allows callers of
`showConsole()` to self-document the usage `requestedScope`.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
  - Fix the bug.
  - Adapt to the clean up for the `options` parameter.
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.showConsole):
  - Adapt to the clean up for the `options` parameter.
* Source/WebInspectorUI/UserInterface/Views/LogContentView.js:
(WI.LogContentView.prototype._showConsoleTab):
  - Adapt to the clean up for the `options` parameter.

Canonical link: https://commits.webkit.org/275143@main

webkit-commit-queue commented:
Committed 275143@main (08a1f3998383): <https://commits.webkit.org/275143@main>

Reviewed commits have been landed. Closing PR #24831 and removing active labels.
